Add Net Present Value Calculation to Costing Framework

bpaul4 opened the pull request:
## Summary/Motivation:
Adds method to calculate the net present value (NPV), which is a better objective than cost of recovery to determine plant economic viability.

## Changes proposed in this PR:
- Add method to calculate NPV from capital costs, operating costs, and revenue

- Operates as an optional cost framework option `build_process_costs([other arguments], calculate_NPV=True, [NPV arguments])` when both fixed and variable operating costs are calculated by referencing cost model variables and adding NPV calculations to the flowsheet costing block
- Also operations as a standalone method `QGESSCostingData.calculate_NPV(total_capital_cost, annual_operating_cost, annual_revenue, cost_year, [NPV_arguments])` that may be attached to any flowsheet block

- Supports a capital expenditure period, and independent cost escalation/inflation of capital costs, operating costs, and revenue via a calculated series present worth factor
- Supports royalties based on revenue or an optional user Expression for royalties
- Supports amortized loan interest based on a debt percentage of capital costs or an optional user Expression for debt
- Supports capital depreciation assuming a declining balance model
- Appropriate method docstrings and tests

## Reviewer's checklist / merge requirements:
- [ ] The head branch (i.e. the "source" of the changes) is not the `main` branch on the PR author's fork
- [ ] Documentation
- [ ] Tests
- [ ] Diagnostic tests for models

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.md file
   at the top level of this directory.

2. I represent I am authorized to make the contributions and grant the license. If my employer has
   rights to intellectual property that includes these contributions, I represent that I have
   received permission to make contributions and grant the required license on behalf of that
   employer.

bpaul4 commented:
I will be adding more tests to cover the different use cases and switches, including the standalone case. In the meantime, please review the current code and tests. The pytests all pass locally.

andrewlee94 commented on src/prommis/uky/costing/ree_plant_capcost.py:
```diff
@@ -2675,3 +2736,672 @@ def costing_upper_bound_eq(c, p):
+        # flowsheet block or costing block
+        b,
+        # required NPV arguments
+        discount_percentage,
```
This is a lot of arguments the user needs to provide to the method. This makes me think it would be better to define these in a `ConfigDict` on the class which the user can then set and play with more easily (and can be better documented), and this method can then access the values from the `ConfigDict`.

bpaul4 replied:
Thank you for the suggestion. I will look up the proper syntax for implementing this approach.

bpaul4 replied:
Would the new `CONFIG` variable belong to the `QGESSCostingData` class and include every argument, so I'd reformulate the other methods as well? I'm looking at code examples of implementations and deciding what I need to add to this file.

andrewlee94 replied:
No. You can overload the config from the parent class so only this class sees the changes.

That is in part why most model classes start with `CONFIG = CONFIG()` - note the `()`. By calling the existing `CONFIG` you are actually making a copy of it for the new class so that any changes you make do not affect the parent (but any changes in the parent will propagate down to the child).

andrewlee94 commented on src/prommis/uky/costing/ree_plant_capcost.py:
```diff
@@ -2675,3 +2736,672 @@ def costing_upper_bound_eq(c, p):
+
+        # check if costing block was passed, and extract necessary components
+        # variables and constraints will be added to the costing block
+        if isinstance(b, FlowsheetCostingBlockData) and b.library == "REE":
```
To help with testing and maintenance, I would suggest breaking all of this input verification out into separate methods that you can unit test.

E.g. all this code for testing the type of the `Block` should be a separate method so that you can quickly test this in isolation by throwing different types of blocks at it (dummied up with and without the expected arguments) so that you can fully test all possible cases without needing to make sure the rest of the code works with the dummy cases.

andrewlee94 commented on src/prommis/uky/costing/ree_plant_capcost.py:
```diff
@@ -2675,3 +2736,672 @@ def costing_upper_bound_eq(c, p):
+            )
+
+        # check capital expenditure arguments
+        if capital_expenditure_percentages is not None:
```
Similarity, with a `ConfigDict` for the arguments, all of this can be broken out into separate verification methods that can be tested in isolation.

andrewlee94 commented on src/prommis/uky/costing/ree_plant_capcost.py:
```diff
@@ -2675,3 +2736,672 @@ def costing_upper_bound_eq(c, p):
+            @b.Constraint()
+            def loan_debt_constraint(c):
+
+                return c.loan_debt == pyunits.convert(
```
Do you need the constraint in these cases? You are adding a lot of simple linear equality constraints to the model which is adding a lot of unnecessary burden on the solver (and writer).

I would suggest for these cases making `c.load_debt` a `Reference` based on the argument provided (noting you will need to move the unit conversion elsewhere).

andrewlee94 replied:
Similarly, you might want to break these sections out as separate sub-methods to make it easier to test them in isolation (so that you can set up a simple Block with the things you need for each branch of the `if/else` and make sure the variables/references and constraints are built correctly).

andrewlee94 commented on src/prommis/uky/costing/tests/test_ree_costing.py:
```diff
@@ -4434,3 +4465,98 @@ def test_REE_costing_recovery_Nonewithtransportcost():
+    assert hasattr(m.fs.costing, "capital_depreciation_declining_balance_percentage")
+
+    # check some NPV results
+    assert m.fs.costing.pv_capital_cost.value == pytest.approx(-14.33727, rel=1e-4)
```
Whilst this verifies the results of the calculations, this does very little to test that the variables and constraint were built as expected. I.e., this is a `component` or `integration` test, but you have no `unit` testing of the code.

andrewlee94 commented on src/prommis/uky/costing/tests/test_ree_costing.py:
```diff
@@ -4434,3 +4465,98 @@ def test_REE_costing_recovery_Nonewithtransportcost():
+
+
+@pytest.mark.component
+def test_REE_costing_NPV_defaults():
```
Building on all of this, you have a lot of `if/else` branches and possible exception paths in the code, but you are not testing any of these here. You need to have a lot more unit testing of these different paths, especially given how complex the code for this model is.

MarcusHolly commented on src/prommis/uky/costing/ree_plant_capcost.py:
```diff
@@ -2669,3 +2858,737 @@ def costing_upper_bound_eq(c, p):
+        # define series present worth factor as an method so it can be called
+
+        def series_present_worth_factor(r, g, N):
+            """
+            Returns expression for series present worth factor.
+            """
```
```suggestion
            """
            Returns expression for series present worth factor where r is the discount rate
            expressed as a decimal, N is the project lifetime, and g is the escalation rate
            (e.g. inflation) expressed as a decimal.
            """
```
It might be worth adding this tidbit here as well for convenience

MarcusHolly commented on src/prommis/uky/costing/ree_plant_capcost.py:
```diff
@@ -2669,3 +2858,737 @@ def costing_upper_bound_eq(c, p):
+        if b.config.debt_expression is None:
+
+            @b.Constraint()
+            def loan_debt_constraint(c):
+                # Debt  = - %debt_charge_of_CAPEX * CAPEX
+
+                return c.loan_debt == pyunits.convert(
+                    pyunits.convert(
+                        c.debt_percentage_of_CAPEX, to_units=pyunits.dimensionless
+                    )
+                    * c.CAPEX,
+                    to_units=c.cost_units,
+                )
```
Should this equation have a negative sign in front?

bpaul4 replied:
```suggestion
            def loan_debt_constraint(c):
                # Debt  = %debt_charge_of_CAPEX * CAPEX

                return c.loan_debt == pyunits.convert(
                    pyunits.convert(
                        c.debt_percentage_of_CAPEX, to_units=pyunits.dimensionless
                    )
                    * c.CAPEX,
                    to_units=c.cost_units,
                )
```

The equation is correct, but there shouldn't be a negative sign in the comment. The debt is taken as a positive value so that the interest owed (a negative cash flow) is `interest_owed = - (payment_per_year * number_of_years - debt)`.

MarcusHolly reviewed (approved):
LGTM

Agfritz commented on src/prommis/uky/costing/ree_plant_capcost.py:
```diff
@@ -2669,3 +2858,739 @@ def costing_upper_bound_eq(c, p):
+        NPV = PV_Revenue - PV_Operating_Cost - PV_Royalties - PV_Capital_Cost
+              - Loan_Interest_Owed - Capital_Depreciation
+
+        For escalating costs, the series present worth factor is modified to
```
I would change this to say "for costs escalating at a constant rate for the project lifetime..." to help distinguish it from capital escalation

Agfritz commented on src/prommis/uky/costing/ree_plant_capcost.py:
```diff
@@ -2669,3 +2858,739 @@ def costing_upper_bound_eq(c, p):
+        owed on loans and value losses due to capital depreciation. The loan
+        interest is calculated as an annual loan payment
+
+        Annual_Loan_Payment = Debt * [ iLoan_% * (1 + iLoan_%)**Nloan ] / [ (1 + iLoan_%)**Nloan - 1 ]
```
I am a little bit confused about why the loan payment is included in the calculation. Typically, the interest is accounted for in the present worth factor. In the later calculations, it appears that you are already converting everything to present value. If this is the case, you do not need to include the loan interest in your summation for NPV (but please let me know if i am misinterpreting something in the formulation).

bpaul4 replied:
I think you correct in your assessment. I understood that the interest should be included as a cost, but if it is already included in capital component then I can remove the loan repayment from the formulation. I'm happy to discuss it further here or offline.

bpaul4 replied:
The case study source provides a few rates - 3% inflation for operating costs, 3% inflation for revenue, 10% discount rate, 3.6% escalation rate for capital costs, and 6% interest rate on loans. Would it correct to use the 6% interest rate as the discount rate when projecting capital expenditure costs, with a 3.6% growth rate? And then 10% (in this scenario) would be the discount rate for operating costs, revenue and royalties, with a 3% growth rate.

Agfritz reviewed (commented):
I have included mostly minor comments but did have one significant concern about the inclusion of loan interest in the final NPV summation. Please let me know if it would be helpful to discuss further.

bpaul4 commented on src/prommis/uky/costing/ree_plant_capcost.py:
```diff
@@ -97,9 +98,151 @@ def custom_REE_plant_currency_units():
+        ConfigValue(
+            default=None,
+            domain=list,
+            description="a list of values that sum to 100 representing how capital costs are spread over a capital expenditure period; for example, an input of [10, 60, 30] is parsed as a 3-year period where capital costs are spread as 10% in year 1, 60% in year 2, and 30% in year 3. The capital period precedes the operating period. Set to None to indicate no expenditure period.",
```
Yes, you are correct. No expenditure period means that 100% of the CAPEX occurs in first operating year t=0. An expenditure period of 1 year (e.g. `capital_expenditure_period=[100,]` would mean that "Year 1" is 100% of the CAPEX and the OPEX starts in the second year.

I'll update the description.

bpaul4 commented on src/prommis/uky/costing/ree_plant_capcost.py:
```diff
@@ -1731,6 +1899,23 @@ def get_fixed_OM_costs(
             )
 
+        # set default values
+        if labor_types is None:
```
The arguments are set using `None` as a guard value so that a new object is created every time the method is called and there is no chance of the object changing on each run.

Agfritz replied:
Sounds good, thank you for clarifying Brandon!

bpaul4 commented:
@Agfritz please re-review this at your convenience and let me know if you have further suggestions to improve the NPV formulation.

The last commit looks odd in the file comparison view but is actually a small change. I added a check

```
if (fixed_OM is False and calculate_NPV is True):  # NPV method invoked with fixed inputs
    self.calculate_NPV(fixed_OM, variable_OM)
else:  # continue on with building the plant costs
```

on line 457 to avoid creating unnecessary model variables and constraints when the user sets fixed values for the CAPEX, OPEX, and revenue to calculate the NPV. All of the code building the plant (overnight) costs, calling the O&M methods, etc is now indented under the `else`, and was otherwise not changed in any way.

Agfritz commented on src/prommis/uky/costing/ree_plant_capcost.py:
```diff
@@ -2669,3 +2869,564 @@ def costing_upper_bound_eq(c, p):
+        )
+
+        b.pv_operating_cost = Var(
+            initialize=-b.CAPEX,
```
Should this be initialized as -b.OPEX? (similar question for initialization of other variables, not sure if it is important here)

bpaul4 replied:
CAPEX is used to initialize the PV values as a rough 1-2 order of magnitude initial guess based on user inputs. I agree that we can update the initial guesses, as there may be cases where CAPEX, OPEX and REVENUE do not have similar values.

Agfritz reviewed (approved):
These changes addressed my concerns. I have added two minor comments to clarify.

bpaul4 commented on src/prommis/uky/costing/ree_plant_capcost.py:
```diff
@@ -2669,3 +2869,564 @@ def costing_upper_bound_eq(c, p):
+            )
+
+        if b.config.royalty_expression is None:
+            # PV_Royalties = - %royalty_charge_of_revenue * REVENUE
```
Yes, that is correct, I will update the comment.

ksbeattie commented:
@andrewlee94, can you re-review this?

andrewlee94 commented on src/prommis/uky/costing/ree_plant_capcost.py:
```diff
@@ -97,9 +98,144 @@ def custom_REE_plant_currency_units():
+            default=3.6,
+            domain=float,
+            description="rate at which capital costs escalate during the "
+            "capital expenditure period. Set to 0 to indicate expenditure is "
```
"Set to 0 to indicate expenditure is spread but there is no cost escalation" implies there is an option for when the expenditure is not spread and there is no cost escalation. I am not sure this makes sense, in which case I would ask if the desired meaning is retained if "expenditure is spread but" is removed from the statement.

andrewlee94 commented on src/prommis/uky/costing/ree_plant_capcost.py:
```diff
@@ -97,9 +98,144 @@ def custom_REE_plant_currency_units():
+        ConfigValue(
+            default=6,
+            domain=float,
+            description="interest rate for capital" "equipment loan repayment.",
```
Remove the  " " from the middle of the doc string (this was caused by `black` removing an unneeded line break).

andrewlee94 commented on src/prommis/uky/costing/ree_plant_capcost.py:
```diff
@@ -97,9 +98,144 @@ def custom_REE_plant_currency_units():
+        ConfigValue(
+            default=3,
+            domain=float,
+            description="inflation rate for operating costs during the "
```
What are the units/basis for this? Given the values, I suspect this is a percentage value which needs to be stated explicitly.

andrewlee94 commented on src/prommis/uky/costing/ree_plant_capcost.py:
```diff
@@ -97,9 +98,144 @@ def custom_REE_plant_currency_units():
+    CONFIG.declare(
+        "revenue_inflation_percentage",
+        ConfigValue(
+            default=3,
```
As above, and also below.

andrewlee94 commented on src/prommis/uky/costing/ree_plant_capcost.py:
```diff
@@ -133,17 +269,10 @@ def build_process_costs(
-            "engineer",
-        ],
-        labor_rate=[27.90, 23.26, 30.29, 24.06, 23.43, 46.82],
+        labor_types=None,
```
Do you actually need all of these argument; they all look like things I just saw in the ConfigDict. If so, this seems like duplication of effect and an opportunity for something to go wrong. How do these arguments get used and populated?

Also, you should be able to remove the `# pylint: disable-next=dangerous-default-value` line above this.

andrewlee94 replied:
My point is that these already exist in the ConfigDict do they not? If that is the case, why create a new local variable for something that is already stored elsewhere.

andrewlee94 commented on src/prommis/uky/costing/ree_plant_capcost.py:
```diff
@@ -274,6 +404,34 @@ def build_process_costs(
+            watertap_block: list of unit model blocks corresponding to watertap models
+            calculate_NPV: True/false flag for calculating net present value (NPV).
+
+            Keyword arguments related to NPV, entered at class instantiation:
```
All of this should already be included in the main class doc string from the config block (and as mentioned above I think all these arguments are just duplicating effort).

andrewlee94 commented on src/prommis/uky/costing/ree_plant_capcost.py:
```diff
@@ -296,648 +454,673 @@ def build_process_costs(
+                self.piping_materials_and_labor_percentage = Param(
+                    mutable=True,
+                    initialize=piping_materials_and_labor_percentage / 100,
+                    doc="Piping, materials and labor",
```
These doc strings do not really describe what this value is. Also, I note you are converting from a percentage value as an input to a decimal value for the parameter. I think this is a bad idea, as it creates a separation between the user input and the actual model which will result in confusion (i.e. a user might go to update one of these manually and expect to provide the value in the same basis they used when declaring the class).

andrewlee94 commented on src/prommis/uky/costing/ree_plant_capcost.py:
```diff
@@ -296,648 +454,673 @@ def build_process_costs(
+                    initialize=value(self.total_BEC),
+                    bounds=(0, 1e4),
+                    doc="Ancillary cost in $MM",
+                    units=getattr(pyunits, "MUSD_" + CE_index_year),
```
Rather than build the units string each time you use it, just build it once up front as a local object and use it wherever it is needed. Otherwise, you are duplicating operations and wasting CPU time.

Also, what happens if the user wants to use Euros or some other currency units?

bpaul4 replied:
See answer below.

andrewlee94 commented on src/prommis/uky/costing/ree_plant_capcost.py:
```diff
@@ -296,648 +454,673 @@ def build_process_costs(
+                self.plant_services_costs = Var(
+                    initialize=value(self.total_BEC),
+                    bounds=(0, 1e4),
+                    doc="Ancillary cost in $MM",
```
Minor comment, but I dislike $MM as this is a very US-centric definition (outside the US a single M indicates millions). Also, there is no need for the currency basis to be dollars - it could well be Euros or GBP, etc.

Thus, I would recommend a more general definition of the units in the doc string.

bpaul4 replied:
Addressing this comment and the one above, we don't have a dictionary of cost index values for non-USD currency units. However, users may add their own custom units by registering them, for example:

```
pyunits.load_definitions_from_strings(
        [ "custom_currency_units= 500/700 * USD_CE500",]
    )
```

where `custom_currency_units` is effectively a CE index value of 700, but really this is currency conversion rate between USD_CE500 and the new units. If users know the conversion, say 1.05 euro to 1 dollar in 2022, they could do something like

```
pyunits.load_definitions_from_strings(
        [ "EUR_2022= 1/1.05 * USD_2022",]
    )
```

We would need to update the code, variables bounds, doc strings, etc. to account for other units besides USD.

andrewlee94 replied:
Actually, a better answer would be to never write units into the doc strings and just store them as units on the object as this gets around the entire issue and avoids duplicating information. We should then make use of the fact that pint supports the concept of `Munits = 1e6*units` to handle the fact that this is in millions of currency units (also as a note for the future, some currencies would want different multipliers due to smaller or larger units of currency).

andrewlee94 commented on src/prommis/uky/costing/ree_plant_capcost.py:
```diff
@@ -296,648 +454,673 @@ def build_process_costs(
+                # buildings cost variables
+                self.buildings_costs = Var(
+                    initialize=value(self.total_BEC),
+                    bounds=(0, 1e4),
```
How reasonable are these bounds, especially given that the basis and units of cost are not well known?

bpaul4 replied:
For millions of USD, the scale of plants should not exceed 10,000 million USD for any of the cost variables so that should be a sufficiently large upper bound. The supported USD cost years have index values between ~350 to 850, so the upper bound works for any cost year.

We would probably want to change or remove the upper bound if we're allowing cost units other than "millions of USD".

andrewlee94 replied:
The CEI values have been steadily increasing over the years however, so eventually they will reach a point where the bound is not reasonable (although that might not happen in a realistic timeframe). There is also the question of people wanting to use non-USD currencies which will have different values.

The real question here is not what is a reasonable value for the bounds, but whether bounds are necessary at all. I.e. given that these values are ultimately based on the sum of costs of individual pieces of equipment, and those costs are in turn based on dimensions of the equipment, these values should be implicitly bounded by the dimensions of the equipment. Adding additional bounds here is essentially adding an arbitrary bound on the dimensions of the equipment that might be tighter than necessary (or alternatively is a bound that will never be active due to tighter constraints elsewhere in the model).

andrewlee94 commented on src/prommis/uky/costing/ree_plant_capcost.py:
```diff
@@ -296,648 +454,673 @@ def build_process_costs(
+                self.contingency_cost_eq = Constraint(rule=contingency_cost_rule)
 
-            self.contingency_cost_eq = Constraint(rule=contingency_cost_rule)
+                def total_installation_cost_rule(self):
```
A minor suggestion, but the `@m.Constraint` decorator generally results in more compact and readable code.

andrewlee94 commented on src/prommis/uky/costing/ree_plant_capcost.py:
```diff
@@ -296,648 +454,673 @@ def build_process_costs(
+            if feed_input is not None:
+                if (
+                    pyunits.get_units(feed_input) == pyunits.dimensionless
+                ):  # assume it's short ton per hour
```
Is this a good assumption? Generally when there is any ambiguity (or in this case an outright unit inconsistency) I generally recommend raising an Exception instead of making an assumption (which is quite likely wrong).

andrewlee94 reviewed (changes requested):
There is a lot to work through here, but here are some comments so far.

bpaul4 commented on src/prommis/uky/costing/ree_plant_capcost.py:
```diff
@@ -97,9 +98,144 @@ def custom_REE_plant_currency_units():
+            "back to their present value. The NETL QGESS recommends setting "
+            "the discount rate as the calculated after-tax weighted average "
+            "cost of ccapital (ATWACC). For the UKy case study, the ATWACC is"
+            "5.77%.",
```
For 1, Users should specify the discount percentage and project lifetime, as these two inputs have an outsized impact on the NPV calculation and are project-specific. I think it better to enforce that users set values for these two inputs and include the comment about the typical value rather than have defaults.

I'll update the doc strings to address 2 and 3. All percentage inputs are assumed to be percentage values (e.g. 5.77 means 5.77%).

andrewlee94 replied:
Regarding 1 then, I would remove the comment about the value for the UKy case study, as it is too-specific for a general tool.

bpaul4 commented on src/prommis/uky/costing/ree_plant_capcost.py:
```diff
@@ -97,9 +98,144 @@ def custom_REE_plant_currency_units():
+        "capital_expenditure_percentages",
+        ConfigValue(
+            default=None,
+            domain=list,
```
I see that `ListOf` supports arguments `itemtype`, `domain`, and `string_lexer`. Is there a way to use `domain` or `string_lexer` to validate something like `ListOf(float, domain=values_sum_to_100)`?

andrewlee94 replied:
I doubt it, but you would need to read the docs.

bpaul4 commented:
@andrewlee94 please let me know if the changes have addressed your comments. I opened https://github.com/prommis/prommis/issues/92 and added a couple of items related to your suggestions.

andrewlee94 commented on src/prommis/uky/costing/ree_plant_capcost.py:
```diff
@@ -110,50 +110,55 @@ class QGESSCostingData(FlowsheetCostingBlockData):
+            description="Value for total capital cost (including equipment, "
+            "installation, and other plant costs); ignored if no value is "
+            "passed. Can be a Var, Param, or Expression with currency units, "
+            "or can specific a cost year in the cost_year argument.",
```
Typo: specify

andrewlee94 commented on src/prommis/uky/costing/ree_plant_capcost.py:
```diff
@@ -475,177 +473,177 @@ def build_process_costs(
                 self.equipment_installation_costs = Var(
                     initialize=value(self.total_BEC),
                     bounds=(0, 1e4),
                     doc="Equipment installation EPCM cost in $MM",
```
Given the change below, I do not think we should keep the "$MM" in the doc string, as it can quite easily be wrong. For determining the units of something, it is far better to directly query the units attribute than to look at a doc string that is likely static.

bpaul4 replied:
I took out all of the docstring and report references to dollars, $, or USD. The method to autodetect reference dictionary units still looks for "$", that's something that will need to be generalized when we do that currency update.

bpaul4 replied:
Also, the report method has a new line to report the currency units.

andrewlee94 reviewed (commented):
I found a typo and a couple of other minor things, but I think my main concerns have either been addressed here or moved to issue #92. I also see that more docs are in a different issue as well.

andrewlee94 commented on src/prommis/uky/costing/ree_plant_capcost.py:
```diff
@@ -1096,162 +1095,153 @@ def initialize_build(*args, **kwargs):
 
         if hasattr(self, "total_plant_cost"):
-            var_dict["Total Plant Cost [$MM]"] = value(self.total_plant_cost)
+            var_dict["Total Plant Cost"] = value(self.total_plant_cost)
```
As a passing note (not for this PR), you can use `as_quantity` (from Pyomo) to get a `pint` Quantity instead (which contains units, and has some useful features for automatically converting the value to a set of display units).

andrewlee94 commented on src/prommis/uky/costing/ree_plant_capcost.py:
```diff
@@ -2533,36 +2516,30 @@ def total_BEC_eq(c):
         if hasattr(b, "total_variable_OM_cost"):
             print(
-                "Total annual variable O&M cost: $%.3f Million"
+                "Total annual variable O&M cost: %.3f"
```
Another minor note, but you might want to consider `fstrings` for this (I find them easier to read).
